t3/utils/flux.py

```diff
@@ -317,35 +327,47 @@ def run_jsr(gas: ct.Solution,
         V (float, optional): The reactor volume in cm^3.
         a_tol (float, optional): The absolute tolerance for the simulation.
         r_tol (float, optional): The relative tolerance for the simulation.
+        tau_tolerance (float, optional): Cantera will solve JSR equations with step() till tau*tolerance time, after that, it will solve them by advance(t_i). This will adress stiff equestion in which small time steps are required.
 
     Returns:
         dict: The T, P, X, and ROP profiles (values) at a specific time.
     """
     V = V or 100
+    num_steps = 500
     profiles = dict()
     stoichiometry = get_rxn_stoichiometry(gas)
     for tau in times:
-        network, reactor = set_jsr(gas=gas, time=tau, composition=composition, T=T, P=P, V=V, a_tol=a_tol, r_tol=r_tol)
         t = 0
-        while t < tau:
-            rops = {spc.name: dict() for spc in gas.species()}
-            t = network.step()
-            if t > tau:
-                network.advance(tau)
-                t = tau
-            cantera_reaction_rops = gas.net_rates_of_progress
-            for spc in gas.species():
-                for i, rxn in enumerate(gas.reactions()):
-                    if stoichiometry[spc.name][i]:
-                        if rxn.equation not in rops[spc.name].keys():
-                            rops[spc.name][rxn.equation] = 0
-                        rops[spc.name][rxn.equation] += cantera_reaction_rops[i] * stoichiometry[spc.name][i]
-            profile = {'P': gas.P, 'T': gas.T, 'X': {s.name: x for s, x in zip(gas.species(), gas.X)}, 'ROPs': rops}
-            if t == tau:
-                profiles[t] = profile
+        dt = tau/num_steps
+        time_steps_array = np.arange(t, tau+dt, dt) #the last number is tau
+        network, reactor = set_jsr(gas=gas, time=tau, composition=composition, T=T, P=P, V=V, a_tol=a_tol, r_tol=r_tol)
+        while t <= tau*tau_tolerance:
+            network.step()
+            #update the system's time
+            t = network.time
+        if t > tau*tau_tolerance:
+            #get the current time of the system after step()
+            #find the closest_bigger time value in time_array, and continue with advance func
+            t = network.time
+            t_array, t_step_index = closest_bigger_number(time_steps_array, t)
+            t = t_array
```

Comment:
maybe instead of these two lines simply do `t, t_step_index = closest_bigger_number(time_steps_array, t)`?

Reply:
was this addressed?